File: spring-boot/jwt/src/main/java/com/example/jwt/auth/application/AuthService.java
package com.example.jwt.auth.application;

import com.example.jwt.auth.application.dto.RefreshTokenRequest;
import com.example.jwt.auth.application.dto.SignInRequest;
import com.example.jwt.auth.application.dto.SignUpRequest;
import com.example.jwt.auth.application.dto.TokenResponse;
import com.example.jwt.auth.exception.DuplicateEmailException;
import com.example.jwt.auth.exception.NotFoundUserException;
import com.example.jwt.auth.exception.NotMatchPasswordException;
import com.example.jwt.user.application.dto.UserResponse;
import com.example.jwt.user.domain.User;
import com.example.jwt.user.domain.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final TokenProvider tokenProvider;
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public TokenResponse signIn(SignInRequest signInRequest) {
        User user = userRepository.findByEmail(signInRequest.getEmail()).orElseThrow(NotFoundUserException::new);
        if (!passwordEncoder.matches(signInRequest.getPassword(), user.getPassword())) {
            throw new NotMatchPasswordException();
        }
        return tokenProvider.create(user);
    }

    public void signOut(Authentication auth) {
        String accessToken = (String) auth.getCredentials();
        tokenProvider.deleteByAccessToken(accessToken);
    }

    @Transactional
    public void signUp(SignUpRequest signUpRequest) {
        if (userRepository.existsByEmail(signUpRequest.getEmail())) {
            throw new DuplicateEmailException();
        }

        userRepository.save(signUpRequest.toUser(passwordEncoder));
    }

    public TokenResponse refresh(RefreshTokenRequest refreshTokenRequest) {
        return tokenProvider.refresh(refreshTokenRequest);
    }

    public UserResponse me(Authentication auth) {
        String email = (String) auth.getPrincipal();
        User user = userRepository.findByEmail(email).orElseThrow(NotFoundUserException::new);
        return UserResponse.of(user);
    }
}

File: task/payhere/src/main/kotlin/com/example/payhere/domain/auth/service/AuthService.kt
package com.example.payhere.domain.auth.service

import com.example.payhere.domain.auth.controller.dto.SignInRequest
import com.example.payhere.domain.auth.controller.dto.SignInResponse
import com.example.payhere.domain.auth.controller.dto.SignUpRequest
import com.example.payhere.domain.auth.exception.SignInFailedException
import com.example.payhere.domain.auth.support.JwtStorage
import com.example.payhere.domain.user.controller.dto.UserDto
import com.example.payhere.domain.user.entity.User
import com.example.payhere.domain.user.entity.mapper.UserMapper
import com.example.payhere.domain.user.entity.repository.UserRepository
import com.example.payhere.domain.user.exception.UserExpiredException
import com.example.payhere.domain.user.exception.UserNotFoundException
import jakarta.transaction.Transactional
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class AuthService(
    private val mapper: UserMapper,
    private val repository: UserRepository,
    private val jwtTokenProvider: JwtTokenProvider,
    private val passwordEncoder: PasswordEncoder
) {
    @Transactional
    fun signIn(signInRequest: SignInRequest): SignInResponse {
        val user = repository
            .findByEmail(signInRequest.email!!)
            .orElseThrow { UserNotFoundException() }
            .apply { if (!passwordEncoder.matches(signInRequest.password!!, password)) throw SignInFailedException() }
            .apply { if (deletedAt != null) throw UserExpiredException() }

        val accessToken = jwtTokenProvider.createToken(user.id.toString())
        return SignInResponse(accessToken = accessToken)
    }

    fun signOut(token: String) {
        JwtStorage.signOut(token)
    }

    @Transactional
    fun signUp(signUpRequest: SignUpRequest) {
        repository.save(
            User(
                email = signUpRequest.email!!,
                password = passwordEncoder.encode(signUpRequest.password!!)
            )
        )
    }

    @Transactional
    fun me(currentUser: User): UserDto {
        return mapper.getDto(currentUser)
    }
}